# Our Week 4 Assignment makes use of the functionality provided by Python's NumPy package.

## 1. Write a Python/NumPy code block that finds the distinct/unique common items between these two arrays:

In [670]:
import numpy as np

In [671]:
a = np.array([1,2,3,2,3,4,3,4,5,6])
b = np.array([7,2,10,2,7,4,9,4,9,8])

## Your output should contain only the distinct overlapping values. For example, if a 2 is found in both array  a  and array  b, your output should contain only one 2 even if array _a_ contains more than one 2 within it.

In [672]:
'''I will get the sorted, common elements in array a and b 
using "intersect1d" array set operation as below.'''
task1_answer = np.intersect1d(a,b)

In [673]:
print(task1_answer) #task1 answer.

[2 4]


## 2. Create the following 5x3 array using knowledge you have of Python's/NumPy's sequencing functionality so that you do not need to explicitly key in every integer value.

In [674]:
myarray = np.arange(1,16)

In [675]:
print(myarray)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15]


In [676]:
myarray2 = np.arange(1,16).reshape(3,5)

In [677]:
print(myarray2)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]]


In [678]:
myarray3 = myarray2.T

In [679]:
print(myarray3) #task2 answer.

[[ 1  6 11]
 [ 2  7 12]
 [ 3  8 13]
 [ 4  9 14]
 [ 5 10 15]]


## 3. You are given the following two arays:

In [680]:
a = np.array([12,5,7,15,3,1,8])
b = np.array([14,6,3,11,19,12,5])

## Write a Python/Numpy code block that removes from array a any items that are also present in array b.

In [681]:
'''I will use "setdiff1d" array set operation to get the array that 
contains elements from array a that are not in b'''
task3_answer = np.setdiff1d(a,b)
print(task3_answer)

[ 1  7  8 15]


## 4. Suppose you have ridden a bicycle from New York City to Key West, Florida. Your bicycle odometer shows the total miles you have travelled thus far, which you make a note of each day with paper and pencil. Your first two entries might be '55' and '120', indicating that you rode your bike 55 miles on day 1 and 65 miles on day 2. Your task is to create a NumPy array wherein you can record the cumulative miles you recorded each day during your trip. Then use your Python skills to show the total miles that you rode each day.  

## Assume the trip took a total of 35 days and included at least 4 non-consecutive days where no cycling was possible due to either weather conditions or personal fatigue/soreness.

## For the remaining 31 days, you would have covered the entire distance between NYC and Key West. Complete your work within your Jupyter Notebook for this assignment and be sure to include a writeup explaining your approach to this problem, including the ways in which you decided to make use of NumPy.

- My approach to this problem:
  1. I googled the distance between NYC and Key West        (https://www.google.com/maps/dir/New+York/Key+West,+Florida+33040/@40.7256193,-73.99434,13.55z/data=!4m14!4m13!1m5!1m1!1s0x89c24fa5d33f083b:0xc80b8f06e177fe62!2m2!1d-74.0059728!2d40.7127753!1m5!1m1!1s0x88d1b134ad952377:0x3fcee92f77463b5e!2m2!1d-81.7799871!2d24.5550593!3e1) and it was 1548 miles by bike.
  2. I will generate the 35 random integer, to simulate the situation where I am recording the miles daily by hand.
  3. I will write code blocks to append the 35 integers to a new list,
     checking the sum of elements should be 1548.
     I will limit the miles per day from 40 to maximum 65, considering 1548miles/31days = 49.93 miles/day.
     Considering 35/4 = 8.75, the maximum days for consecutive bike ride will be 7 days,
     I take a break every 8th day.  
  4. I will convert the new list into ndarray using "np.asarray" method.
  5. I will use "np.diff" array method to create the array that shows the total miles I rode each day.

In [682]:
import random #random module is built-in, but I made sure to import it.
miles = np.random.randint(40, 65, size=35) #drawing random 35 integers from 40 to 65 to populate "dailiy_miles" list. 
daily_miles = miles.tolist() #converting "daily_miles" array into a list.
odometer = [] #I will append the cumulative miles to this list.
cum_miles = 0 #setting the initial value for the cumulative miles as 0.
day = [n for n in range(35)] #n is value from 0 to 34, as there are 35 days.
for n in day: #this will iterate from n=0 until n reaches 34.
    if n > 0 and n % 8 == 0: #I will take a break every 8th day, as described above.
        odometer.append(cum_miles) #odometer will have the same value as the previous element on every 8th day.
    else:
        cum_miles += daily_miles[n] #cumulative miles will be increased by the daily_miles each day otherwise.
        if n != 34 and cum_miles <=1548:
            odometer.append(cum_miles)
        else:
            odometer.append(1548)
        

In [683]:
print(odometer)
type(odometer) #checking the object type.

[60, 105, 147, 189, 234, 289, 332, 388, 388, 444, 498, 549, 612, 655, 697, 756, 756, 809, 853, 896, 957, 1015, 1065, 1107, 1107, 1152, 1203, 1249, 1312, 1358, 1402, 1460, 1460, 1510, 1548]


list

In [684]:
odometer_array = np.asarray(odometer) #converting the list into ndarray.
print(odometer_array)
type(odometer_array) #checking the object type.

[  60  105  147  189  234  289  332  388  388  444  498  549  612  655
  697  756  756  809  853  896  957 1015 1065 1107 1107 1152 1203 1249
 1312 1358 1402 1460 1460 1510 1548]


numpy.ndarray

In [685]:
'''I will calculate the daily miles that I rode from the "odometer_array" 
that shows accumulated miles from day 1 to day 35 by using "np.diff" method.'''
daily_miles = np.diff(odometer_array)
print(daily_miles) #task 4 answer.

[45 42 42 45 55 43 56  0 56 54 51 63 43 42 59  0 53 44 43 61 58 50 42  0
 45 51 46 63 46 44 58  0 50 38]


## Task 5.1~5.4

In [686]:
import pandas
import numpy as np

# supress scientific notation
np.set_printoptions(suppress = True)

# read the data file from GitHub and convert to a NumPy ndarray object
nyc_water = pandas.read_csv('https://raw.githubusercontent.com/jtopor/DAV-5400/master/Week4/Water_Consumption_In_The_New_York_City.csv').as_matrix()

# verify you have an ndarray object
type(nyc_water)

C:\Users\Minyeong\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: FutureWarning: Method .as_matrix will be removed in a future version. Use .values instead.
  


numpy.ndarray

In [687]:
nyc_water

array([[   1979. , 7102100. ,    1512. ,     213. ],
       [   1980. , 7071639. ,    1506. ,     213. ],
       [   1981. , 7089241. ,    1309. ,     185. ],
       [   1982. , 7109105. ,    1382. ,     194. ],
       [   1983. , 7181224. ,    1424. ,     198. ],
       [   1984. , 7234514. ,    1465. ,     203. ],
       [   1985. , 7274054. ,    1326. ,     182. ],
       [   1986. , 7319246. ,    1351. ,     185. ],
       [   1987. , 7342476. ,    1447. ,     197. ],
       [   1988. , 7353719. ,    1484. ,     202. ],
       [   1989. , 7344175. ,    1402. ,     191. ],
       [   1990. , 7335650. ,    1424. ,     194. ],
       [   1991. , 7374501. ,    1469. ,     199. ],
       [   1992. , 7428944. ,    1369. ,     184. ],
       [   1993. , 7506166. ,    1368.5,     182. ],
       [   1994. , 7570458. ,    1357.7,     179. ],
       [   1995. , 7633040. ,    1325.7,     174. ],
       [   1996. , 7697812. ,    1297.9,     169. ],
       [   1997. , 7773443. ,    1205.5,     1

## task 5-1. What is the maximum yearly NYC consumption of water in millions of gallons per day?

In [688]:
'''I will find the maximum value from the third column of nyc_water by using
"max" array statistical methods as below:'''
task5_1 = nyc_water[:, 2]
np.max(task5_1)

1512.0

- task 5-1 answer: Above output shows that maximum yearly NYC consumption of water in millions of gallons per day was 1,512, as the third column is the corresoponding column for NYC Consumption(Million gallons per day).

## task 5-2. How many calendar years are represented within this data set? NumPy's shape command is one way to find out.

In [689]:
task5_2 = np.shape(nyc_water)
print(task5_2)
task5_2[0]

(39, 4)


39

- task 5-2 answer: Above output shows that this data set has 39 rows and 4 columns. Each row represents each year, so 39 calendar years are represented within this data set.

## task 5-3. What is the mean and standard deviation of the per capita daily water consumption?

In [690]:
task5_3 = nyc_water[:, 3] #slicing the last(4th) column
print(np.mean(task5_3)) #calculating the mean along the column
print(np.std(task5_3)) #calculating the standard deviation along the column

160.30769230769232
31.51894330370624


- task 5-3 answer: Above output shows that the last column, which corresponds to the per capita daily water consumption, has the mean value 160.30769231, and the standard deviation 31.5189433.

## task 5-4. What is the increase or decrease in population from year to year? Use NumPy's 'diff' function to create an array of differences and save that to a variable called "pop_diff", then print that variable to the screen.

In [691]:
'''First, I will sort the nyec_water by the "year"(0th column)
and then calculate the difference between the row for each column as below.
Later, I found out that the array was sorted by year by default, but it was good to learn 
how to sort by each column.'''
sorted_year = nyc_water[nyc_water[:,0].argsort()] #sorting the nyc_water array by the "year"(0th column)
print(sorted_year)

[[   1979.  7102100.     1512.      213. ]
 [   1980.  7071639.     1506.      213. ]
 [   1981.  7089241.     1309.      185. ]
 [   1982.  7109105.     1382.      194. ]
 [   1983.  7181224.     1424.      198. ]
 [   1984.  7234514.     1465.      203. ]
 [   1985.  7274054.     1326.      182. ]
 [   1986.  7319246.     1351.      185. ]
 [   1987.  7342476.     1447.      197. ]
 [   1988.  7353719.     1484.      202. ]
 [   1989.  7344175.     1402.      191. ]
 [   1990.  7335650.     1424.      194. ]
 [   1991.  7374501.     1469.      199. ]
 [   1992.  7428944.     1369.      184. ]
 [   1993.  7506166.     1368.5     182. ]
 [   1994.  7570458.     1357.7     179. ]
 [   1995.  7633040.     1325.7     174. ]
 [   1996.  7697812.     1297.9     169. ]
 [   1997.  7773443.     1205.5     155. ]
 [   1998.  7858259.     1219.5     155. ]
 [   1999.  7947660.     1237.2     156. ]
 [   2000.  8008278.     1240.4     155. ]
 [   2001.  8024963.5    1184.      148. ]
 [   2002. 

In [692]:
'''The code block below will show increase or decrease 
in population from year to year in the second column'''
task5_4 = np.diff(sorted_year, axis=0) #calculating the difference between each row for each column.
print(task5_4)
pop_diff=task5_4[:, 1] #slicing the second column to show the difference of "New York City Population" from year to year.
print(pop_diff) #task5-4 answer.

[[     1.  -30461.      -6.       0. ]
 [     1.   17602.    -197.     -28. ]
 [     1.   19864.      73.       9. ]
 [     1.   72119.      42.       4. ]
 [     1.   53290.      41.       5. ]
 [     1.   39540.    -139.     -21. ]
 [     1.   45192.      25.       3. ]
 [     1.   23230.      96.      12. ]
 [     1.   11243.      37.       5. ]
 [     1.   -9544.     -82.     -11. ]
 [     1.   -8525.      22.       3. ]
 [     1.   38851.      45.       5. ]
 [     1.   54443.    -100.     -15. ]
 [     1.   77222.      -0.5     -2. ]
 [     1.   64292.     -10.8     -3. ]
 [     1.   62582.     -32.      -5. ]
 [     1.   64772.     -27.8     -5. ]
 [     1.   75631.     -92.4    -14. ]
 [     1.   84816.      14.       0. ]
 [     1.   89401.      17.7      1. ]
 [     1.   60618.       3.2     -1. ]
 [     1.   16685.5    -56.4     -7. ]
 [     1.   16685.5    -48.4     -7. ]
 [     1.   16685.5    -41.9     -5. ]
 [     1.   16685.5      5.8      0. ]
 [     1.   16685.5     3